src/main/lib/osu-file-parser/OsuParser.ts

```diff
@@ -1,125 +1,345 @@
     const songTable = new Map<ResourceID, Song>();
     const audioTable = new Map<ResourceID, AudioSource>();
-    const imageTable = new Map<ResourceID, ImageSource>();
-
-    for (let i = 0; i < dirs.length; i++) {
-      const subDirPath = path.join(dir, dirs[i]);
 
-      const files = await getFiles(subDirPath, "osu");
-      for (let j = 0; j < files.length; j++) {
-        if (update !== undefined) {
-          update(i + 1, dirs.length, files[j]);
-        }
+    // NOTE: Images can only be fetched from the .osu file AFAIK.
+    //       You should parse those dynamically when needed.
+    const imageTable = new Map<ResourceID, ImageSource>();
 
-        const file = path.join(subDirPath, files[j]);
-        const audioSrc = await OsuFile.getProp(file, "AudioFilename");
+    // HACK: Used by prototype overrides above
+    db.pos = 0;
 
-        if (audioSrc.isError) {
-          continue;
-        }
+    const db_version = db.read_u32();
+    console.debug("db version:", db_version);
+    if (db_version < 20170222) {
+      return fail("osu!.db is too old, please update the game.");
+    }
 
-        const audioID = path.join(dirs[i], audioSrc.value);
-        let s: Song | undefined = undefined;
+    db.read_u32(); // folder count
+    db.read_u8();
+    db.read_u64(); // timestamp
 
-        if (audioID === lastAudioID) {
-          s = lastAudioSong;
-        } else {
-          const a = audioTable.get(audioID);
+    const player_name = db.read_string().trim();
+    console.debug("player name:", player_name);
 
-          if (a !== undefined) {
-            s = songTable.get(a.songID);
-          }
+    const nb_beatmaps = db.read_u32();
+    console.debug("nb beatmaps:", nb_beatmaps);
 
-          if (s !== undefined) {
-            lastAudioID = audioID;
-            lastAudioSong = s;
-          }
+    let last_audio_filepath = "";
+    for (let i = 0; i < nb_beatmaps; i++) {
+      try {
+        if (db_version < 20191107) {
+          // https://osu.ppy.sh/home/changelog/stable40/20191107.2
+          db.read_u32();
         }
 
-        if (s !== undefined) {
-          const diff = await OsuFile.getProp(file, "Version");
+        const song: Song = {
+          audio: "",
+          osuFile: "",
+          path: "",
+          ctime: "",
+          dateAdded: "",
+          title: "",
+          artist: "",
+          creator: "",
+          bpm: [],
+          duration: 0,
+          diffs: []
+        };
+
+        song.artist = db.read_string().trim();
+        song.artistUnicode = db.read_string().trim();
+        song.title = db.read_string().trim();
+        song.titleUnicode = db.read_string().trim();
+        song.creator = db.read_string().trim();
+
+        // NOTE: I'm being lazy here, and only loading the first diff for a given audio file.
+        //       This is wrong, especially if diffs of a set aren't right next to each other in the db.
+        const diff_name = db.read_string();
```

Comment:
All difficulties shall be loaded. Searching requires it

Reply:
fixed